tests/zfs-tests/tests/functional/cli_root/zfs_send/zfs_send_holds.ksh

```diff
@@ -0,0 +1,176 @@
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
+# Use is subject to license terms.
+#
+
+#
+# Copyright (c) 2012, 2016 by Delphix. All rights reserved.
+# Copyright (c) 2018 Datto Inc. All rights reserved.
+#
+
+. $STF_SUITE/tests/functional/cli_root/cli_common.kshlib
+. $STF_SUITE/tests/functional/cli_root/zfs_send/zfs_send.cfg
+
+#
+# DESCRIPTION:
+#	Verify 'zfs send' can send dataset holds.
+#	Verify 'zfs receive' can receive dataset holds.
+#
+# STRATEGY:
+#	1. Create a snapshot.
+#	2. Create a full send stream with the fs, including holds.
+#	3. Receive the send stream and verify the data integrity.
+#	4. Fill in fs with some new data.
+#	5. Create an incremental send stream with the fs, including holds.
+#	6. Receive the incremental send stream and verify the data integrity.
+#	7. Verify the holds have been received as expected.
+#	8. Create an incremental snap with no holds, and send that with -h.
+#	9. Confirm the snapshot was received as expected.
+#
+
+verify_runnable "both"
+
+function cleanup
+{
+	log_must zfs release hold1-1 $init_snap > /dev/null
+	log_must zfs release hold1-2 $init_snap > /dev/null
+	log_must zfs release hold1-1 $recv_snap > /dev/null
+	log_must zfs release hold1-2 $recv_snap > /dev/null
+	log_must zfs release hold2-1 $inc_snap > /dev/null
+	log_must zfs release hold2-1 $recv_inc_snap  >/dev/null
+	for snap in $init_snap $inc_snap $recv_snap $recv_inc_snap; do
+                snapexists $snap && log_must zfs destroy -f $snap
+        done
+
+	datasetexists $recv_root && log_must zfs destroy -Rf $recv_root
+
+	for file in $full_bkup $inc_bkup $init_data $inc_data $inc_bkup2
+	do
+		[[ -e $file ]] && log_must rm -f $file
+	done
+
+	[[ -e $file ]] && log_must rm -f $inc_data2
+
+	[[ -d $TESTDIR1 ]] && log_must rm -rf $TESTDIR1
+
+}
+
+#
+# Check if hold(s) exist on the specified dataset.
+# Fail test if requested holds are not found.
+#
+function hold_check #<dataset>..<hold1>..<holdn>
+{
+	typeset -i count=0
+	typeset dataset=""
+
+	for hold in $@; do
+		if [[ $count == 0 ]]; then
+			dataset=$hold
+		else
+			zfs holds $dataset |grep -q $hold
+			if (($? != 0)); then
+				log_fail "dataset hold $hold was not found"
+				break
+			fi
+		fi
+		(( count +=1 ))
+	done
+}
+
+log_assert "Verify 'zfs send/recv' can send and receive dataset holds."
+log_onexit cleanup
+
+init_snap=$TESTPOOL/$TESTFS@init_snap
+inc_snap=$TESTPOOL/$TESTFS@inc_snap
+inc_snap2=$TESTPOOL/$TESTFS@inc_snap2
+full_bkup=/var/tmp/fullbkup.$$
```

Comment:
nit: use `$TEST_BASE_DIR` instead of hardcoded paths.

Reply:
nit: please clean these up after the test as well, thanks